File: web_console/backend/services/camera_discovery.py
"""绑定指定有线网口的海康 SADP / ONVIF WS-Discovery 发现。

本模块只负责二层/UDP 发现与报文解析，不读取或修改主机网络配置。SADP 固件未
广播的服务端口会带 ``*_port_inferred`` 标记，交由页面明确提示并允许人工覆盖。
"""
from __future__ import annotations

import html
import ipaddress
import re
import select
import socket
import struct
import time
import uuid
import xml.etree.ElementTree as ET
from typing import Any, Dict, Iterable, List
from urllib.parse import unquote, urlsplit


DISCOVERY_GROUP = "239.255.255.250"
DISCOVERY_PORT = 37020
SO_BINDTODEVICE = getattr(socket, "SO_BINDTODEVICE", 25)
_DEVICE_RE = re.compile(r"^[A-Za-z0-9_.:-]{1,15}$")
_MAC_RE = re.compile(r"^[0-9a-f]{12}$")


class CameraDiscoveryError(RuntimeError):
    pass


def _validate_device(device: str) -> str:
    clean = str(device).strip()
    if not _DEVICE_RE.fullmatch(clean):
        raise CameraDiscoveryError("非法的网卡名称")
    try:
        socket.if_nametoindex(clean)
    except OSError as exc:
        raise CameraDiscoveryError(f"网卡 {clean} 不存在") from exc
    return clean


def _normalize_mac(value: str) -> str:
    compact = "".join(char.lower() for char in value if char.lower() in "0123456789abcdef")
    if not _MAC_RE.fullmatch(compact):
        return value.strip().lower()
    return ":".join(compact[index:index + 2] for index in range(0, 12, 2))


def _local_name(tag: str) -> str:
    return tag.rsplit("}", 1)[-1].rsplit(":", 1)[-1].lower()


def _xml_values(payload: str) -> Dict[str, List[str]]:
    values: Dict[str, List[str]] = {}
    try:
        root = ET.fromstring(payload)
    except ET.ParseError:
        # 个别旧固件会在 XML 末尾附加 NUL；清理后再尝试一次。
        try:
            root = ET.fromstring(payload.rstrip("\x00\r\n "))
        except ET.ParseError:
            return values
    for element in root.iter():
        text = html.unescape((element.text or "").strip())
        if text:
            values.setdefault(_local_name(element.tag), []).append(text)
    return values


def _first(values: Dict[str, List[str]], names: Iterable[str]) -> str:
    for name in names:
        items = values.get(name.lower())
        if items:
            return items[0]
    return ""


def _port(value: str) -> int | None:
    try:
        port = int(value)
    except (TypeError, ValueError):
        return None
    return port if 1 <= port <= 65535 else None


def _prefix_from_mask(mask: str) -> int | None:
    if not mask:
        return None
    try:
        return ipaddress.IPv4Network(f"0.0.0.0/{mask}").prefixlen
    except (ValueError, ipaddress.NetmaskValueError):
        return None


def _onvif_model(scopes: str) -> str:
    marker = "onvif://www.onvif.org/hardware/"
    for scope in scopes.split():
        if scope.lower().startswith(marker):
            return unquote(scope[len(marker):]).strip()
    return ""


def parse_discovery_payload(payload: bytes | str, sender_ip: str = "") -> Dict[str, Any] | None:
    """解析单个 SADP/WS-Discovery 响应；无法识别时返回 ``None``。"""
    if isinstance(payload, bytes):
        text = payload.decode("utf-8", errors="replace")
    else:
        text = payload
    values = _xml_values(text)
    if not values:
        return None
    lower_text = text.lower()
    is_response = any(key in values for key in (
        "ipv4address", "devicedescription", "probematch", "xaddrs",
    )) or "probematch" in lower_text
    if not is_response:
        return None

    camera_ip = _first(values, ("ipv4address", "ipaddress", "ipv4"))
    http_port = _port(_first(values, ("httpport",)))
    xaddrs = _first(values, ("xaddrs", "xaddr"))
    if xaddrs:
        for item in xaddrs.split():
            try:
                parsed = urlsplit(item)
                candidate = ipaddress.IPv4Address(parsed.hostname or "")
            except ValueError:
                continue
            camera_ip = str(candidate)
            if http_port is None and parsed.port:
                http_port = parsed.port
            break
    if not camera_ip:
        camera_ip = sender_ip
    try:
        camera_ip = str(ipaddress.IPv4Address(camera_ip))
    except ValueError:
        return None

    model = _first(values, (
        "devicedescription", "devicemodel", "model", "devicetype",
    ))
    scopes = " ".join(values.get("scopes", []))
    if not model:
        model = _onvif_model(scopes)
    subnet_mask = _first(values, ("ipv4subnetmask", "subnetmask"))
    rtsp_port = _port(_first(values, ("rtspport",)))
    result = {
        "ip": camera_ip,
        "mac": _normalize_mac(_first(values, ("mac", "macaddress"))),
        "model": model,
        "serial": _first(values, ("devicesn", "serialnumber")),
        "subnet_mask": subnet_mask,
        "prefix_length": _prefix_from_mask(subnet_mask),
        "gateway": _first(values, ("ipv4gateway", "gateway", "defaultgateway")),
        "http_port": http_port or 80,
        "rtsp_port": rtsp_port or 554,
        "http_port_inferred": http_port is None,
        "rtsp_port_inferred": rtsp_port is None,
        "hikvision": (
            "hikvision" in lower_text
            or "devicedescription" in values
            or _first(values, ("types",)).lower() == "inquiry"
        ),
        "source": "onvif" if "xaddrs" in values else "sadp",
    }
    return result


def _merge_camera(target: Dict[str, Any], incoming: Dict[str, Any]) -> None:
    for key in ("ip", "mac", "model", "serial", "subnet_mask", "prefix_length", "gateway"):
        if target.get(key) in (None, "") and incoming.get(key) not in (None, ""):
            target[key] = incoming[key]
    for port_name in ("http", "rtsp"):
        port = f"{port_name}_port"
        inferred = f"{port_name}_port_inferred"
        if target.get(port) is None or (target.get(inferred) and not incoming.get(inferred)):
            target[port] = incoming.get(port)
            target[inferred] = incoming.get(inferred, False)
    target["hikvision"] = bool(target.get("hikvision") or incoming.get("hikvision"))
    sources = set(str(target.get("source", "")).split("+")) | set(str(incoming.get("source", "")).split("+"))
    target["source"] = "+".join(sorted(item for item in sources if item))


class CameraDiscovery:
    """在一个明确指定的物理网口上搜索摄像头。"""

    @staticmethod
    def discover(device: str, timeout_seconds: float = 3.0) -> List[Dict[str, Any]]:
        clean = _validate_device(device)
        timeout_seconds = max(0.2, min(float(timeout_seconds), 15.0))
        interface_index = socket.if_nametoindex(clean)
        probe_id = str(uuid.uuid4())
        sadp = (
            '<?xml version="1.0" encoding="utf-8"?>'
            f"<Probe><Uuid>{probe_id}</Uuid><Types>inquiry</Types></Probe>"
        ).encode()
        onvif = (
            '<?xml version="1.0" encoding="UTF-8"?>'
            '<e:Envelope xmlns:e="http://www.w3.org/2003/05/soap-envelope" '
            'xmlns:w="http://schemas.xmlsoap.org/ws/2004/08/addressing" '
            'xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery" '
            'xmlns:dn="http://www.onvif.org/ver10/network/wsdl">'
            f"<e:Header><w:MessageID>uuid:{probe_id}</w:MessageID>"
            '<w:To e:mustUnderstand="true">urn:schemas-xmlsoap-org:ws:2005:04:discovery</w:To>'
            '<w:Action e:mustUnderstand="true">http://schemas.xmlsoap.org/ws/2005/04/discovery/Probe</w:Action>'
            '</e:Header><e:Body><d:Probe><d:Types>dn:NetworkVideoTransmitter</d:Types>'
            '</d:Probe></e:Body></e:Envelope>'
        ).encode()

        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as discovery_socket:
            discovery_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            if hasattr(socket, "SO_REUSEPORT"):
                try:
                    discovery_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
                except OSError:
                    pass
            discovery_socket.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
            try:
                discovery_socket.setsockopt(
                    socket.SOL_SOCKET, SO_BINDTODEVICE, clean.encode() + b"\0",
                )
            except OSError as exc:
                raise CameraDiscoveryError(f"无法绑定网卡 {clean}：{exc}") from exc
            try:
                discovery_socket.bind(("0.0.0.0", DISCOVERY_PORT))
            except OSError:
                discovery_socket.bind(("0.0.0.0", 0))
            membership = struct.pack(
                "=4s4sI", socket.inet_aton(DISCOVERY_GROUP), b"\0" * 4, interface_index,
            )
            try:
                discovery_socket.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, membership)
                discovery_socket.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_IF, membership)
            except OSError:
                # 没有 IPv4 的新网口仍可通过定向广播发 SADP，组播失败不应阻止搜索。
                pass

            sent = 0
            for payload, destination in (
                (sadp, (DISCOVERY_GROUP, DISCOVERY_PORT)),
                (sadp, ("255.255.255.255", DISCOVERY_PORT)),
                (onvif, (DISCOVERY_GROUP, DISCOVERY_PORT)),
            ):
                try:
                    discovery_socket.sendto(payload, destination)
                    sent += 1
                except OSError:
                    continue
            if sent == 0:
                raise CameraDiscoveryError(f"无法从网卡 {clean} 发送发现报文，请检查物理 Link 和权限")

            cameras: Dict[str, Dict[str, Any]] = {}
            deadline = time.monotonic() + timeout_seconds
            while time.monotonic() < deadline:
                readable, _, _ = select.select(
                    [discovery_socket], [], [], max(0.0, deadline - time.monotonic()),
                )
                if not readable:
                    break
                try:
                    payload, sender = discovery_socket.recvfrom(65535)
                except OSError:
                    continue
                camera = parse_discovery_payload(payload, sender[0])
                if camera is None:
                    continue
                key = camera.get("mac") or camera["ip"]
                existing_key = next((
                    current_key for current_key, current in cameras.items()
                    if current.get("ip") == camera.get("ip")
                    or (
                        current.get("mac") and camera.get("mac")
                        and current.get("mac") == camera.get("mac")
                    )
                ), None)
                if existing_key is not None:
                    _merge_camera(cameras[existing_key], camera)
                else:
                    cameras[key] = camera
            return sorted(cameras.values(), key=lambda item: ipaddress.IPv4Address(item["ip"]))

File: first_net_config/src/camera_discovery.c
#define _GNU_SOURCE
#define _POSIX_C_SOURCE 200809L

#include "camera_discovery.h"
#include "cli_io.h"
#include "ipv4_utils.h"

#include <arpa/inet.h>
#include <ctype.h>
#include <errno.h>
#include <ifaddrs.h>
#include <net/if.h>
#include <netinet/in.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <sys/select.h>
#include <sys/socket.h>
#include <time.h>
#include <unistd.h>

#ifndef SO_BINDTODEVICE
#define SO_BINDTODEVICE 25
#endif

#define SADP_PORT 37020
#define ONVIF_PORT 3702
#define DISCOVERY_GROUP "239.255.255.250"

static const char *SADP_PROBE =
    "<?xml version=\"1.0\" encoding=\"utf-8\"?>"
    "<Probe><Uuid>FNC-CAMERA-DISCOVERY</Uuid><Types>inquiry</Types></Probe>";

static const char *ONVIF_PROBE =
    "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
    "<e:Envelope xmlns:e=\"http://www.w3.org/2003/05/soap-envelope\" "
    "xmlns:w=\"http://schemas.xmlsoap.org/ws/2004/08/addressing\" "
    "xmlns:d=\"http://schemas.xmlsoap.org/ws/2005/04/discovery\" "
    "xmlns:dn=\"http://www.onvif.org/ver10/network/wsdl\">"
    "<e:Header><w:MessageID>uuid:FNC-CAMERA-DISCOVERY</w:MessageID>"
    "<w:To e:mustUnderstand=\"true\">urn:schemas-xmlsoap-org:ws:2005:04:discovery</w:To>"
    "<w:Action e:mustUnderstand=\"true\">http://schemas.xmlsoap.org/ws/2005/04/discovery/Probe</w:Action>"
    "</e:Header><e:Body><d:Probe><d:Types>dn:NetworkVideoTransmitter</d:Types>"
    "</d:Probe></e:Body></e:Envelope>";

static bool tag_name_matches(const char *start, const char *end,
                             const char *wanted)
{
    const char *local = start;
    const char *cursor;
    size_t length;

    for (cursor = start; cursor < end; ++cursor)
    {
        if (*cursor == ':')
        {
            local = cursor + 1;
        }
        if (isspace((unsigned char)*cursor) || *cursor == '/' || *cursor == '>')
        {
            end = cursor;
            break;
        }
    }
    length = (size_t)(end - local);
    return strlen(wanted) == length && strncasecmp(local, wanted, length) == 0;
}

static bool xml_has_element(const char *xml, const char *wanted)
{
    const char *cursor = xml;

    if (!xml || !wanted)
    {
        return false;
    }
    while ((cursor = strchr(cursor, '<')) != NULL)
    {
        const char *name = cursor + 1;
        const char *tag_end = strchr(name, '>');

        if (!tag_end)
        {
            return false;
        }
        if (*name != '/' && *name != '?' && *name != '!' &&
            tag_name_matches(name, tag_end, wanted))
        {
            return true;
        }
        cursor = tag_end + 1;
    }
    return false;
}

static bool xml_value(const char *xml, const char *wanted,
                      char *out, size_t out_size)
{
    const char *cursor = xml;

    if (!xml || !wanted || !out || out_size == 0)
    {
        return false;
    }
    out[0] = '\0';
    while ((cursor = strchr(cursor, '<')) != NULL)
    {
        const char *name = cursor + 1;
        const char *tag_end = strchr(name, '>');
        const char *value_end;
        size_t length;

        if (!tag_end)
        {
            return false;
        }
        if (*name == '/' || *name == '?' || *name == '!' ||
            !tag_name_matches(name, tag_end, wanted))
        {
            cursor = tag_end + 1;
            continue;
        }
        value_end = strchr(tag_end + 1, '<');
        if (!value_end)
        {
            return false;
        }
        length = (size_t)(value_end - (tag_end + 1));
        while (length > 0 && isspace((unsigned char)tag_end[1]))
        {
            ++tag_end;
            --length;
        }
        while (length > 0 && isspace((unsigned char)tag_end[length]))
        {
            --length;
        }
        if (length >= out_size)
        {
            length = out_size - 1;
        }
        memcpy(out, tag_end + 1, length);
        out[length] = '\0';
        return length > 0;
    }
    return false;
}

static void first_xml_value(const char *xml, const char *const names[],
                            char *out, size_t out_size)
{
    out[0] = '\0';
    for (int index = 0; names[index]; ++index)
    {
        if (xml_value(xml, names[index], out, out_size))
        {
            return;
        }
    }
}

static void ip_from_xaddrs(const char *xaddrs, char *ip, size_t ip_size,
                           int *port)
{
    const char *scheme;
    const char *host;
    const char *end;
    char candidate[64];

    if (!xaddrs || !(scheme = strstr(xaddrs, "://")))
    {
        return;
    }
    host = scheme + 3;
    end = host;
    while (*end && *end != ':' && *end != '/' && !isspace((unsigned char)*end))
    {
        ++end;
    }
    if ((size_t)(end - host) >= sizeof(candidate))
    {
        return;
    }
    memcpy(candidate, host, (size_t)(end - host));
    candidate[end - host] = '\0';
    if (valid_ipv4(candidate))
    {
        size_t length = strlen(candidate);
        if (length < ip_size)
        {
            memcpy(ip, candidate, length + 1);
        }
    }
    if (*end == ':')
    {
        long parsed = strtol(end + 1, NULL, 10);
        if (parsed >= 1 && parsed <= 65535)
        {
            *port = (int)parsed;
        }
    }
}

static int parsed_port(const char *text, int fallback)
{
    char *end = NULL;
    long value;

    if (!text || !*text)
    {
        return fallback;
    }
    errno = 0;
    value = strtol(text, &end, 10);
    return errno == 0 && end != text && *end == '\0' &&
                   value >= 1 && value <= 65535
               ? (int)value
               : fallback;
}

int parse_camera_discovery_payload(const char *payload, const char *sender_ip,
                                   DiscoveredCamera *camera)
{
    static const char *const ip_names[] = {
        "IPv4Address", "IPAddress", "IPv4", NULL};
    static const char *const mask_names[] = {
        "IPv4SubnetMask", "SubnetMask", NULL};
    static const char *const gateway_names[] = {
        "IPv4Gateway", "Gateway", "DefaultGateway", NULL};
    static const char *const model_names[] = {
        "DeviceDescription", "DeviceModel", "Model", "DeviceType", NULL};
    static const char *const serial_names[] = {
        "DeviceSN", "SerialNumber", NULL};
    static const char *const mac_names[] = {"MAC", "MACAddress", NULL};
    char xaddrs[512] = {0};
    char port_text[32] = {0};

    if (!payload || !camera)
    {
        return 0;
    }

    /*
     * 本机发送的 SADP/ONVIF 探测请求也可能经组播或广播回到 socket。
     * 请求只有 Probe，真正的发现响应使用 ProbeMatch。不能把请求报文的
     * 本机源 IP 当成摄像头地址。
     */
    if (xml_has_element(payload, "Probe") &&
        !xml_has_element(payload, "ProbeMatch"))
    {
        return 0;
    }

    memset(camera, 0, sizeof(*camera));
    camera->http_port = 80;
    camera->rtsp_port = 554;
    camera->prefix = -1;
    first_xml_value(payload, ip_names, camera->ip, sizeof(camera->ip));
    (void)xml_value(payload, "XAddrs", xaddrs, sizeof(xaddrs));
    if (camera->ip[0] == '\0' && xaddrs[0] != '\0')
    {
        ip_from_xaddrs(xaddrs, camera->ip, sizeof(camera->ip),
                       &camera->http_port);
    }
    if (camera->ip[0] == '\0' && valid_ipv4(sender_ip))
    {
        snprintf(camera->ip, sizeof(camera->ip), "%s", sender_ip);
    }
    if (!valid_ipv4(camera->ip))
    {
        return 0;
    }

    first_xml_value(payload, mask_names, camera->subnet_mask,
                    sizeof(camera->subnet_mask));
    first_xml_value(payload, gateway_names, camera->gateway,
                    sizeof(camera->gateway));
    first_xml_value(payload, model_names, camera->model,
                    sizeof(camera->model));
    first_xml_value(payload, serial_names, camera->serial,
                    sizeof(camera->serial));
    first_xml_value(payload, mac_names, camera->mac, sizeof(camera->mac));
    if (camera->subnet_mask[0])
    {
        camera->prefix = netmask_to_prefix(camera->subnet_mask);
    }
    if (xml_value(payload, "HttpPort", port_text, sizeof(port_text)))
    {
        camera->http_port = parsed_port(port_text, 80);
    }
    if (xml_value(payload, "RtspPort", port_text, sizeof(port_text)))
    {
        camera->rtsp_port = parsed_port(port_text, 554);
    }
    return 1;
}

static void merge_camera(DiscoveredCamera *target,
                         const DiscoveredCamera *incoming)
{
#define COPY_IF_EMPTY(field)                                                   \
    do                                                                         \
    {                                                                          \
        if (target->field[0] == '\0' && incoming->field[0] != '\0')            \
            snprintf(target->field, sizeof(target->field), "%s",              \
                     incoming->field);                                         \
    } while (0)
    COPY_IF_EMPTY(subnet_mask);
    COPY_IF_EMPTY(gateway);
    COPY_IF_EMPTY(mac);
    COPY_IF_EMPTY(model);
    COPY_IF_EMPTY(serial);
#undef COPY_IF_EMPTY
    if (target->prefix < 0 && incoming->prefix >= 0)
    {
        target->prefix = incoming->prefix;
    }
}

static bool ipv4_is_local_address(const char *ip)
{
    struct ifaddrs *interfaces = NULL;
    bool found = false;

    if (!valid_ipv4(ip) || getifaddrs(&interfaces) != 0)
    {
        return false;
    }
    for (struct ifaddrs *item = interfaces;
         item != NULL; item = item->ifa_next)
    {
        char address[INET_ADDRSTRLEN];

        if (!item->ifa_addr || item->ifa_addr->sa_family != AF_INET ||
            !inet_ntop(AF_INET,
                       &((struct sockaddr_in *)item->ifa_addr)->sin_addr,
                       address, (socklen_t)sizeof(address)))
        {
            continue;
        }
        if (strcmp(address, ip) == 0)
        {
            found = true;
            break;
        }
    }
    freeifaddrs(interfaces);
    return found;
}

static int send_probe(int socket_fd, const char *payload,
                      const char *address, int port)
{
    struct sockaddr_in destination = {0};

    destination.sin_family = AF_INET;
    destination.sin_port = htons((uint16_t)port);
    if (inet_pton(AF_INET, address, &destination.sin_addr) != 1)
    {
        return -1;
    }
    return sendto(socket_fd, payload, strlen(payload), 0,
                  (struct sockaddr *)&destination, sizeof(destination)) >= 0
               ? 0
               : -1;
}

int discover_cameras(const char *iface,
                     DiscoveredCamera *cameras,
                     size_t capacity,
                     int timeout_seconds)
{
    int socket_fd;
    int enabled = 1;
    unsigned int ifindex;
    struct sockaddr_in local = {0};
    struct ip_mreqn multicast = {0};
    int sent = 0;
    int found = 0;
    time_t deadline;

    if (!iface || !cameras || capacity == 0 ||
        (ifindex = if_nametoindex(iface)) == 0)
    {
        return -1;
    }
    memset(cameras, 0, capacity * sizeof(*cameras));
    socket_fd = socket(AF_INET, SOCK_DGRAM, 0);
    if (socket_fd < 0)
    {
        return -1;
    }
    (void)setsockopt(socket_fd, SOL_SOCKET, SO_REUSEADDR,
                     &enabled, sizeof(enabled));
    (void)setsockopt(socket_fd, SOL_SOCKET, SO_BROADCAST,
                     &enabled, sizeof(enabled));
    if (setsockopt(socket_fd, SOL_SOCKET, SO_BINDTODEVICE,
                   iface, (socklen_t)(strlen(iface) + 1)) != 0)
    {
        close(socket_fd);
        return -1;
    }
    local.sin_family = AF_INET;
    local.sin_addr.s_addr = htonl(INADDR_ANY);
    local.sin_port = htons(SADP_PORT);
    if (bind(socket_fd, (struct sockaddr *)&local, sizeof(local)) != 0)
    {
        local.sin_port = 0;
        if (bind(socket_fd, (struct sockaddr *)&local, sizeof(local)) != 0)
        {
            close(socket_fd);
            return -1;
        }
    }
    multicast.imr_ifindex = (int)ifindex;
    (void)inet_pton(AF_INET, DISCOVERY_GROUP, &multicast.imr_multiaddr);
    (void)setsockopt(socket_fd, IPPROTO_IP, IP_MULTICAST_IF,
                     &multicast, sizeof(multicast));
    (void)setsockopt(socket_fd, IPPROTO_IP, IP_ADD_MEMBERSHIP,
                     &multicast, sizeof(multicast));

    sent += send_probe(socket_fd, SADP_PROBE, DISCOVERY_GROUP, SADP_PORT) == 0;
    sent += send_probe(socket_fd, SADP_PROBE, "255.255.255.255", SADP_PORT) == 0;
    sent += send_probe(socket_fd, ONVIF_PROBE, DISCOVERY_GROUP, ONVIF_PORT) == 0;
    if (sent == 0)
    {
        close(socket_fd);
        return -1;
    }

    timeout_seconds = timeout_seconds < 1 ? 1 :
                      timeout_seconds > 15 ? 15 : timeout_seconds;
    deadline = time(NULL) + timeout_seconds;
    while (time(NULL) < deadline)
    {
        fd_set read_set;
        struct timeval timeout = {.tv_sec = 1, .tv_usec = 0};
        char payload[65536];
        struct sockaddr_in sender = {0};
        socklen_t sender_size = sizeof(sender);
        ssize_t size;
        char sender_ip[INET_ADDRSTRLEN] = {0};
        DiscoveredCamera parsed;
        int existing = -1;

        FD_ZERO(&read_set);
        FD_SET(socket_fd, &read_set);
        if (select(socket_fd + 1, &read_set, NULL, NULL, &timeout) <= 0)
        {
            continue;
        }
        size = recvfrom(socket_fd, payload, sizeof(payload) - 1, 0,
                        (struct sockaddr *)&sender, &sender_size);
        if (size <= 0)
        {
            continue;
        }
        payload[size] = '\0';
        (void)inet_ntop(AF_INET, &sender.sin_addr, sender_ip,
                        (socklen_t)sizeof(sender_ip));
        if (ipv4_is_local_address(sender_ip))
        {
            continue;
        }
        if (!parse_camera_discovery_payload(payload, sender_ip, &parsed))
        {
            continue;
        }
        for (int index = 0; index < found; ++index)
        {
            if (strcmp(cameras[index].ip, parsed.ip) == 0 ||
                (cameras[index].mac[0] && parsed.mac[0] &&
                 strcasecmp(cameras[index].mac, parsed.mac) == 0))
            {
                existing = index;
                break;
            }
        }
        if (existing >= 0)
        {
            merge_camera(&cameras[existing], &parsed);
        }
        else if ((size_t)found < capacity)
        {
            cameras[found++] = parsed;
        }
    }
    close(socket_fd);
    return found;
}

int choose_or_enter_camera(const char *iface, DiscoveredCamera *camera)
{
    DiscoveredCamera cameras[MAX_DISCOVERED_CAMERAS];
    int count;

    printf("\n正在通过 %s 搜索海康 SADP / ONVIF 摄像头...\n", iface);
    count = discover_cameras(iface, cameras, MAX_DISCOVERED_CAMERAS, 4);
    if (count < 0)
    {
        printf("[提醒] 摄像头搜索失败，请检查 Link、root 权限或端口占用。\n");
        count = 0;
    }
    if (count > 0)
    {
        printf("\n发现摄像头：\n");
        for (int index = 0; index < count; ++index)
        {
            printf("  %d. IP:%-15s 掩码:%-15s MAC:%-20s 型号:%s\n",
                   index + 1, cameras[index].ip,
                   cameras[index].subnet_mask[0] ? cameras[index].subnet_mask : "未广播",
                   cameras[index].mac[0] ? cameras[index].mac : "未广播",
                   cameras[index].model[0] ? cameras[index].model : "未广播");
        }
        printf("  %d. 手动填写\n", count + 1);
        {
            int choice = read_int("请选择摄像头: ", 1, count + 1);
            if (choice <= count)
            {
                *camera = cameras[choice - 1];
                return 1;
            }
        }
    }
    else
    {
        printf("没有发现摄像头。部分固件或交换机可能不转发发现报文，可手动填写。\n");
    }

    memset(camera, 0, sizeof(*camera));
    camera->prefix = -1;
    camera->http_port = 80;
    camera->rtsp_port = 554;
    read_ipv4_required("摄像头 IPv4 地址: ", camera->ip, sizeof(camera->ip));
    return 1;
}
